File: add_strings.py
def addStrings(num1,num2):
    
    #inizializzo le lunghezze

    len1 = len(num1)
    len2 = len(num2)

    max_len=0

    if len1 > len2:         #verifico quale sia la stringa più lunga
        max_len = len1
    else:
        max_len = len2

    i = len1-1              #partendo dalla fine, devo sottrarre un valore, range -> (0-len-1)
    j = len2 -1
    k = 0

    sum_elem,n1,n2,carry=0,0,0,0   #inizializzo, somme e riporto
    l=[]

    while i>=0 or j>=0 or carry:   #mentre sono nel range delle stringhe
        if i>=0:
            n1 = int(num1[i])     #se i è ancora nel range, n1 assume il valore della stringa in pos 1, castata in int
        else:
            n1 = 0               #altrimenti n1 viene settato a 0
        if j>=0:                 #stesso per n2
            n2 = int(num2[j])
        else:
            n2 = 0

        sum_elem = n1 + n2 + carry          #sommo n1,n2, tenendo conto del riporto
        carry = sum_elem //10               #calcolo il riporto con mod(10)
        l.append(str(sum_elem % 10))        #inserisco nella lista la somma castata

        if i >= 0:                          #decremento se i e j sono ancora positivi, per rimanere nel range
            i-=1
        if j >= 0:
            j-=1
        k+=1

    s = "".join(l[::-1])                    #inserisco i caratteri nella mia stringa con join
                                            #faccio attenzione ad inserire i caratteri partendo dalla fine della stringa
        
    return s        
    

num1="11"
num2="123"
print(addStrings(num1,num2))

File: add_strings.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* addStrings(char* num1, char* num2) {
    
    //inizializzo lunghezze, contatori e variabili per calcolare le somme
    int len1 = strlen(num1);
    int len2 = strlen(num2);
    int k=0;
    int i=len1-1;
    int j=len2-1;
    int carry=0;
    int n1=0;
    int n2=0;
    
    int max_len=0;      

    if(len1 > len2){            //controllo la stringa più lunga 
        max_len = len1 +1;
    }
    else{
        max_len = len2 +1;
    }

    char* strSum;                       //alloco memoria
    strSum = malloc((max_len+1)*sizeof(char)); 

    while(i>=0 || j>=0 || carry){       //faccio le somme, mentre i contatori sono ancora nel range della len, gestendo anche l'ultimo riporto

        if(i>=0){               //mentre l'indice i è positivo (in range len(num1)) faccio la conversione
            n1 = num1[i] -'0';
        }
        else
            n1 = 0;
        
        if(j>=0){               //stessa cosa per num2
            n2 = num2[j] - '0';
        }
        else
            n2 = 0;

        int sum = n1 + n2+ carry;       //sommo i valori più il riporto
        carry = (sum) / 10;             //calcolo il riporto 
        strSum[k] = (sum % 10) + '0';  // inserisco nell'array la somma castata in carattere
        
        if(i >= 0)
            i--;
        
        if(j >= 0)
            j--;
        
        k++; 
    }

    strSum[k] = '\0';
    
    int start=0;
    int end=k-1;
    
    //giro i caratteri della stringa, sono partito dalla fine della stringa per calcolare la somma
    
    while(start<end){           
        char temp = strSum[start];
        strSum[start] = strSum[end];
        strSum[end] = temp;
        start++;
        end--;
    }

    return strSum;
    }

int main(){
    char num1[]="11";
    char num2[]="123";


    char* result;
    result = addStrings(num1,num2);
    printf("%s",result);
    free(result);
}   
